packager/mux/mux_test.go

```diff
@@ -0,0 +1,210 @@
+			testURL:       "$HOST/priv/doc?fetch=$FETCH&sign=$SIGN%2A\\",
+			expectHandler: "signer",
+			expectParams:  map[string]string{},
+		}, {
+			testName:      "Signer - with path, empty",
+			testURL:       "$HOST/priv/doc/",
+			expectHandler: "signer",
+			expectParams:  map[string]string{"signURL": ""},
+		}, {
+			testName:      "Signer -  with path, regular",
+			testURL:       "$HOST/priv/doc/$FETCH",
+			expectHandler: "signer",
+			expectParams:  map[string]string{"signURL": "$FETCH"},
+		}, {
+			testName:      "Signer - with path, escaping",
+			testURL:       "$HOST/priv/doc/$FETCH%2A\\",
+			expectHandler: "signer",
+			expectParams:  map[string]string{"signURL": "$FETCH%2A%5C"},
+		}, {
+			testName:      "Signer - with path and query, regular",
+			testURL:       "$HOST/priv/doc/$FETCH?amp=1",
+			expectHandler: "signer",
+			expectParams:  map[string]string{"signURL": "$FETCH?amp=1"},
+		}, {
+			testName:      "Signer - with path and query, escaping",
+			testURL:       "$HOST/priv/doc/$FETCH%2A\\?amp=1%2A\\",
+			expectHandler: "signer",
+			expectParams:  map[string]string{"signURL": "$FETCH%2A%5C?amp=1%2A\\"},
+		}, {
+			testName:      "Cert - empty",
+			testURL:       "$HOST/amppkg/cert/",
+			expectHandler: "cert",
+			expectParams:  map[string]string{"certName": ""},
+		}, {
+			testName:      "Cert - regular",
+			testURL:       "$HOST/amppkg/cert/$CERT",
+			expectHandler: "cert",
+			expectParams:  map[string]string{"certName": "$CERT"},
+		}, {
+			testName:      "Cert - escaping",
+			testURL:       "$HOST/amppkg/cert/$CERT%2A\\",
+			expectHandler: "cert",
+			expectParams:  map[string]string{"certName": "$CERT*\\"},
+		}, {
+			testName:      "ValidityMap - regular",
+			testURL:       "$HOST/amppkg/validity",
+			expectHandler: "validityMap",
+			expectParams:  map[string]string{},
+		}, {
+			testName:      "Healthz - regular",
+			testURL:       "$HOST/healthz",
+			expectHandler: "healthz",
+			expectParams:  map[string]string{},
+		},
+	}
+	for _, tt := range templateTests {
+		testName := tt.testName
+		t.Run(testName, func(t *testing.T) {
+			mocks := map[string](*mockedHandler){"signer": &mockedHandler{}, "healthz": &mockedHandler{}, "cert": &mockedHandler{}, "validityMap": &mockedHandler{}}
+			mux := New(mocks["cert"], mocks["signer"], mocks["validityMap"], mocks["healthz"])
+
+			// expand template URL and expectParams.
+			tt.testURL = expand(tt.testURL)
+			for v, k := range tt.expectParams {
+				tt.expectParams[v] = expand(k)
+			}
+
+			// Set expectation.
+			expectMockedHandler := mocks[tt.expectHandler]
+			expectMockedHandler.On("ServeHTTP", tt.expectParams)
+
+			// Run.
+			actualResp := pkgt.Get(t, mux, tt.testURL)
+
+			// Expect no errors.
+			assert.Equal(t, 200, actualResp.StatusCode, "No error expected: %#v", actualResp)
+
+			// Expect the right call to the right handler, and no calls to the rest.
+			for _, mockedHandler := range mocks {
```

Comment:
Optionally, put this right after you define the mocks, using defer. e.g.

```
defer func() {
  for _, mockedHandler ... {
    ...
  }
}()
```

I generally like this pattern for cleanup because it deals gracefully with early returns and panics. (Similar to `finally`, `__exit__`, and RAII from other languages.)

In this particular case it doesn't make much of a difference because you have no early returns, but it still provides some peace of mind right up front that the cleanup will be handled later.

Reply:
Cool, will do!